File: config/config-annotation-processor/src/test/java/ru/tinkoff/kora/config/annotation/processor/ConfigSourceAnnotationTest.java
package ru.tinkoff.kora.config.annotation.processor;

import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.config.common.Config;
import ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor;
import ru.tinkoff.kora.config.common.factory.MapConfigFactory;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

public class ConfigSourceAnnotationTest extends AbstractConfigTest {

    @Test
    public void testConfigSourceGeneratesConfigExtractor() {
        var extractor = this.compileConfig(List.of(), """
            @ru.tinkoff.kora.config.common.annotation.ConfigSource("test.path")
            public interface TestConfig {
              int value();
            }
            """);

        assertThat(extractor.extract(MapConfigFactory.fromMap(Map.of("value", 42)).root()))
            .isEqualTo(newObject("$TestConfig_ConfigValueExtractor$TestConfig_Impl", 42));
    }

    @Test
    public void testConfigSourceGeneratesModule() throws NoSuchMethodException {
        this.compileConfig(List.of(), """
            @ru.tinkoff.kora.config.common.annotation.ConfigSource("test.path")
            public interface TestConfig {
              int value();
            }
            """);

        var moduleClass = this.compileResult.loadClass("TestConfigModule");
        assertThat(moduleClass)
            .isNotNull()
            .isInterface()
            .hasMethods("testConfig");

        var method = moduleClass.getMethod("testConfig", Config.class, ConfigValueExtractor.class);
        assertThat(method).isNotNull();
        assertThat(method.getReturnType()).isEqualTo(this.compileResult.loadClass("TestConfig"));
        assertThat(method.isDefault()).isTrue();
    }

}

File: config/config-symbol-processor/src/test/kotlin/ru/tinkoff/kora/config/symbol/processor/ConfigSourceAnnotationTest.kt
package ru.tinkoff.kora.config.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.config.common.Config
import ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor
import ru.tinkoff.kora.config.common.factory.MapConfigFactory

class ConfigSourceAnnotationTest : AbstractConfigTest() {
    @Test
    fun testConfigSourceGeneratesConfigExtractor() {
        val extractor = compileConfig(
            listOf<Any>(), """
            @ConfigSource("test.path")
            interface TestConfig {
              fun value() : Int
            }
            
            """.trimIndent()
        )
        assertThat(extractor!!.extract(MapConfigFactory.fromMap(mapOf("value" to 42)).root()))
            .isEqualTo(new("\$TestConfig_ConfigValueExtractor\$TestConfig_Impl", 42))
    }

    @Test
    fun testConfigSourceGeneratesModule() {
        compileConfig(
            listOf<Any>(), """
            @ConfigSource("test.path")
            interface TestConfig {
              fun value() : Int
            }
            
            """.trimIndent()
        )
        val moduleClass = compileResult.loadClass("TestConfigModule")
        assertThat(moduleClass)
            .isNotNull()
            .isInterface()
            .hasMethods("testConfig")

        val method = moduleClass.getMethod("testConfig", Config::class.java, ConfigValueExtractor::class.java)
        assertThat(method).isNotNull()
        assertThat(method.returnType).isEqualTo(compileResult.loadClass("TestConfig"))
        assertThat(method.isDefault).isTrue()
    }
}
